### ISSUE TITLE
improve open/close ring at DisconnectingPoints

### ISSUE THREAD

**Issue by nesnoj** [openego/eDisGo#23]
In Dingo, opening/closing DisconnectingPoints removes/adds the 2 lines from the DP to the adjacent nodes. In eDisGo we better use an attribute of DP to store its state. Subsequently, we must consider this attribute in the pyPSA export.

Additionally, consider Circuit Breakers during import. Currently, those are neglected and two things can happen because of that
1. If CB is closed: line is imported as normal line and we have a real ring (which is closed)
2. If CB is open: the according line isn't imported in eDsiGo

**Comment by nesnoj** [comment]
I stumbled upon this topic again while implementing connection of MV genos (check if there's a MVDisconnectingPoint on the line...etc). I'd like to avoid to implement something which is deleted anyway so I'll solve this now.

One question @gplssm @birgits: Except in the importer, I couldn't find no place where u make use of the `MVDisconnectingPoint`s (open, close, ...), do you? Hence, the grid stability tests are done in closed state, right? This is not what we want, the DPs must be open (normal operation condition) when doing the PF...

I will try to switch to the new approach (see above) and change your code accordingly including opening them..

**Comment by birgits** [comment]
No, the rings are open. I think the disconnecting points are not taken into consideration at all right now.

**Comment by gplssm** [comment]
@nesnoj and @gplssm decided to change representation of disconnection points. I'd like to outline quickly here

1. Switch disconnectors are located in/at LVStation. DisconnectionPoint won't exist anymore
2. The switch disconnector is an own class associated by the LV station that contains the switch disconnector
3. DisconnectionPoints/switch disconnectors have to be relocated after import of Ding0 data
4. A new attribute is introduced at MVGrid class: `switch_disconnector_stations`. This is a generator returning all LV stations that have a switch disconnector
5. The class SwitchDisconnector has the same attributes/methods as the DisconnetionPoint

**Questions**

1. How could the ring be opened without removing an edge or a node?

**TODOs**

- [ ]  Create class SwitchDisconnector
- [ ] Adapt the importer: include all edges (such that the ring would be "closable"), equip LV station with SwitchDisconnector when instantianted
- [ ] Provide a method for load orientated switch disconnector positioning
- [ ] Provide a method at the MVGrid to open all switch disconnectors
- [ ] Adapt the PyPSA-interface

**Comment by gplssm** [comment]
Nice ideas above, but doesn't work :disappointed_relieved: Due to characteristic of graph data structures, we cannot simply open a ring without removing an edge. In order to split a ring, at least a node or an edge has to be removed (where the first option removes both adjacent edges). 

@nesnoj I would like to have a quick call with you tomorrow.

**Comment by nesnoj** [comment]
> Nice ideas above, but doesn't work :disappointed_relieved: Due to characteristic of graph data structures, we cannot simply open a ring without removing an edge

Sure, the graph does not exactly reflect the real grid when the DPs are open, but we can easily consider the DPs in all export functions (e.g. do not export the branch or add an additional DP graph node at the DP station which is not connected to the DP station point).

Task: If we assign a DP to an LV station (or BranchTee!!!), the position is ambiguous -> we have to store the side of the node where the ring is open..

I'm in two minds here since both options (store as attribute or as graph node) have strong advantages & disadvantages - I tend to use the graph node option.. :confused: 

I'll give you a buzz later today as soon as I'm available..

**Comment by gplssm** [comment]
Updates after call with @nesnoj 

* No switch disconnectors during import: line segments that are removed when in circuit breakers of Ding0 grid topology are opened are replaced by one line that closes the ring. Switch disconnectors in eDisGo data structure will be introduced after connection of generators
* Switch disconnector is represented by a non-functional node plus a short line that connects respective LV station with the new node (the non-functional) and is removed in case of opened ring
* The switch disconnector's location in eDisGo is determined again after generators (and in far future loads) are connected to the grid

TODOs from above remain so far.

**Comment by gplssm** [comment]
Detailed and ordered list of TODOs

- [ ] Adapt `import_from_Ding0()` to build grid topology with closed MV rings
- [ ] Adapt grid import data validation not to complain about different number of lines and circuit breakers
- [ ] Eventually adapt `to_pypsa()` (actually, when the short line segment that represents this switch disconnector, is removed from the graph, `to_pypsa()` should work as well)
- [ ] Provide function to position switch disconnecter optimally (provide optionally the case which determines location of the switch disconnecter {feedin,load}). Additionally, provide option to specify list of potential nodes for location of switch disconnecter (i.e. all LV stations on the ring)
- [ ] Add switch disconnecter instance accordingly to next LV station
- [ ] Provide functionality to open/close switch disconnecter for this class
- [ ] Provide method `{open, close}_switch_disconnecters()` to open/close all SDs in the MV grid

**Comment by nesnoj** [comment]
r u on this right now?

**Comment by gplssm** [comment]
> r u on this right now?

I'll drop you a line, when I changed the importer to return closed MV rings without CBs/SDs

**Comment by gplssm** [comment]
Two additional points popped up

1. Actually the line's capacity should be considered determining the optimal switch disconnecter position
2. According to consultancy by @jochenbuehler switch disconnecters are located in LV stations

**Comment by nesnoj** [comment]
> 1. Actually the line's capacity should be considered determining the optimal switch disconnecter position

If I got this right, we perform the following steps:
1. Load grid
2. Import new genos (or loads)
3. Set SD positions
4. Do reinforcement

Hence, the position of the SDs affect the reinforcement measures which are performed after determining the SDs' positions and not vice versa. Thus, the cables after step 4 should be able to carry the power and an additional re-location of SD won't be required...(there might be a small error due to line loading, though). I do not see the point in your idea, could you please provide more details?

**Comment by gplssm** [comment]
You're right and I agree on the order of steps we perform!

And true, there won't no any issues due to line overloading or voltage band violations after grid reinforcement. But, the location of the switch disconnecter may would be different if we would consider the equipment characteristics in step 3. 

Anyhow, I would be fine neglecting it. I think the deviation is small.

**Comment by nesnoj** [comment]
> But, the location of the switch disconnecter may would be different if we would consider the equipment characteristics in step 3.

I assume you mean the slightly different power flow due to the voltage differences which are neglected in the current SD positioning method (which does not involve Q / voltage drops)? If so, I agree.

>  I would be fine neglecting it.

:+1:

**Comment by gplssm** [comment]
I think we're still talking about different things... :smirk: 

Anyhow, I keep it as it is and explain it to you when we meet next time.

**Comment by gplssm** [comment]
Most of the stuff is done in e383f65fea71864db9abf7c1b4f24a8e17344a6c. The only thing missing is that switch disconnecters are always located in LV stations. Hence, we have to exclude other node from list of possible SD locations. This is postponed to v0.0.2.

**Comment by nesnoj** [comment]
Thank you, latenightworker! :+1:

**Comment by birgits** [comment]
When we do grid expansion in the MV grid lines can now only be disconnected at LV stations in order to be able to also operate this part of the grid as a closed ring. Therefore all LV stations need to have disconnecting points and we need to add the information, that disconnecting the line after 2/3 of the line length and connecting the LV station directly to the MV busbar creates a new open ring.
